cpp/src/arrow/type.cc

```diff
@@ -2090,17 +2090,42 @@ Status SchemaBuilder::AreCompatible(const std::vector<std::shared_ptr<Schema>>&
   return Merge(schemas, policy).status();
 }
 
+std::vector<std::shared_ptr<Field>> make_fields(
+    const std::initializer_list<std::pair<std::string, std::shared_ptr<DataType>>>&
+        init_list) {
+  std::vector<std::shared_ptr<Field>> fields;
+  fields.reserve(init_list.size());
+  for (const auto& pair : init_list) {
+    fields.push_back(std::make_shared<Field>(pair.first, pair.second));
```

Comment:
Since we're in c++17, we can use the more readable structured binding:
```suggestion
  for (const auto& [name, type] : init_list) {
    fields.push_back(field(name, type));
```